File: battleship/game/seagrid.cpp
#include "seagrid.h"
#include <assert.h>


SeaGrid::SeaGrid(unsigned short side)
{
	m_gridSize = side * side;
	m_grid = new SQUARE[m_gridSize];
}


SeaGrid::~SeaGrid(void)
{
	delete m_grid;
}


bool SeaGrid::IsSquaresAvailable(int squares[], int squares_len)
{
	for (int i=0; i < squares_len; i++)
	{
		assert (squares[i] < m_gridSize);
		if (m_grid[squares[i]] != AVAILABLE)
			return false;
	}
	return true;
}


/* "not in array" helper function */
inline bool not_in(int value, int array[], int array_len)
{
	for (int i=0; i < array_len; i++)
	{
		if (value == array[i])
			return false;
	}
	return true;
}


void SeaGrid::OccupySquares(int squares[], int squares_len)
{
	assert (IsSquaresAvailable(squares, squares_len));

	for (int i = 0; i < squares_len; i++)
	{
		assert (m_grid[i] == AVAILABLE);
		m_grid[i] = OCCUPIED;

		// horizontal margin
		int pos = i - 1;  // left
		if (i % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			if (not_in(pos, squares, squares_len))
			{
				assert (m_grid[pos] != OCCUPIED);
				m_grid[pos] = BLOCKED;
			}
		}
		pos = i + 1;  // right
		if (pos % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			if (not_in(pos, squares, squares_len))
			{
				assert (m_grid[pos] != OCCUPIED);
				m_grid[pos] = BLOCKED;
			}
		}

		// vertical margin
		pos = i - m_sideSize;  // up
		if (pos >= 0 && pos < m_gridSize)
		{
			if (not_in(pos, squares, squares_len))
			{
				assert (m_grid[pos] != OCCUPIED);
				m_grid[pos] = BLOCKED;
			}
		}
		pos = i + m_sideSize;  // down
		if (pos >= 0 && pos < m_gridSize)
		{
			if (not_in(pos, squares, squares_len))
			{
				assert (m_grid[pos] != OCCUPIED);
				m_grid[pos] = BLOCKED;
			}
		}

		// diagonal margin
		pos = i - m_sideSize + 1;  // upper-right
		if (pos % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			assert (m_grid[pos] != OCCUPIED);
			m_grid[pos] = BLOCKED;
		}
		pos = i - m_sideSize - 1;  // upper-left
		if (i % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			assert (m_grid[pos] != OCCUPIED);
			m_grid[pos] = BLOCKED;
		}
		pos = i + m_sideSize + 1;  // lower-right
		if (pos % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			assert (m_grid[pos] != OCCUPIED);
			m_grid[pos] = BLOCKED;
		}
		pos = i + m_sideSize - 1;  // lower-left
		if (i % m_sideSize != 0 && pos >= 0 && pos < m_gridSize)
		{
			assert (m_grid[pos] != OCCUPIED);
			m_grid[pos] = BLOCKED;
		}
	}
}

File: battleship/seagrid.py
"""
This module purpose is to help track of player's fleets.
It incapsulates implementation of tasks like fleet setup, player shots,
validation of actopns, and provides status information.
This is typical "model" in MVC terminology.
"""
import math

def is_squares_available(sea, indices):
	for i in indices:
		if sea[i] != ' ':
			return False
	return True

def occupy_squares(sea, indices):
	S = int(math.sqrt(len(sea)))
	assert S**2 == len(sea)
	for i in indices:
		assert sea[i] == ' '
		sea[i] = '#'
		# horizontal margin
		if i % S != 0 and i-1 >= 0 and i-1 < len(sea) and i-1 not in indices:
			assert sea[i-1] != '#'
			sea[i-1] = '.'  # left
		if (i+1) % S != 0 and i+1 >= 0 and i+1 < len(sea) and i+1 not in indices:
			assert sea[i+1] != '#'
			sea[i+1] = '.'  # right
		# vertical margin
		if i-S >= 0 and i-S < len(sea) and i-S not in indices:
			assert sea[i-S] != '#'
			sea[i-S] = '.'  # up
		if i+S >= 0 and i+S < len(sea) and i+S not in indices:
			assert sea[i+S] != '#'
			sea[i+S] = '.'  # down
		# diagonal margin
		if (i-S+1) % S != 0 and i-S+1 >= 0 and i-S+1 < len(sea):
			assert sea[i-S+1] != '#'
			sea[i-S+1] = '.'  # upper-right
		if i % S != 0 and i-S-1 >= 0 and i-S-1 < len(sea):
			assert sea[i-S-1] != '#'
			sea[i-S-1] = '.'  # upper-left
		if (i+S+1) % S != 0 and i+S+1 >= 0 and i+S+1 < len(sea):
			assert sea[i+S+1] != '#'
			sea[i+S+1] = '.'  # lower-right
		if i % S != 0 and i+S-1 >= 0 and i+S-1 < len(sea):
			assert sea[i+S-1] != '#'
			sea[i+S-1] = '.'  # lower-left

def shoot_square(sea, pos):
	sea_side = int(math.sqrt(len(sea)))
	assert sea_side**2 == len(sea)
	index = pos[1] * sea_side + pos[0]
	if sea[index] == '#':
		return True
	return False
